### ISSUE TITLE
gRPC deframer complexity

### ISSUE THREAD

**Issue by trustin** [line/armeria#2952]
Is it only myself who finds `ArmeriaMessageDeframer` more complex than necessary? I think we can simplify it in the long term by writing a building block like the following:

Write a Reactive Streams `Processor` which decodes a stream of `HttpObject`s into another stream. It is different from `FilteredStreamMessage` in that it allows decoding M `HttpObject`s into N objects. This will provide some extension point like Netty's `ByteToMessageDecoder` does.

On top of this building block, we could rewrite `ArmeriaDeframer` with much less state management.

It may be possible to do something similar for `ArmeriaMessageFramer`, but I didn't think much about it yet.

**Comment by minwoox** [comment]
> Write a Reactive Streams Processor which decodes a stream of HttpObjects into another stream.

We can use that for multi-part as well. 😉

**Comment by jrhee17** [comment]
Thinking of taking a stab at this (unless someone is already working on it)

**Comment by ikhoon** [comment]
I was about to start this issue ;-)
No matter who start it, let's start to talk about API design.

**Comment by jrhee17** [comment]
I know you're already working on multipart stuff, so I'll look for another issue.
I'll try to pop in for reviews instead 😄

**Comment by ikhoon** [comment]
I think we can think about two points.

1) Reorganize the existing elements.
In `MultipartEncoder` case, it does not need byte-level operations. It only needs a mapping function to publish N sequence.
```java
class StreamMessageProcessor<T, U> implements StreamMessage<U>, Processor<T, U> {
   /* Flushes all buffered data (I'm not sure this operation is required) */
   void flush();
   void flushBefore(T element);
   void flushUntil(T element);

   List<U> process(List<T> elements);
}
2) Deframe chunked bytes to messages. 
ArmeriaMessageDeframer and MultipartDecoder(#2894) decodes HTTP payloads to concretes messages.
```java
class HttpDataDeframer<T> implements StreamMessageProcessor<HttpData, T> {
    // Needs virtual buffer to access bytes across the `HttpData`s
    
    List<U> process(HttpData data) {
       virtualBuffer.add(data);
       returns process()
    }
    abstract List<T> process(); 

}

class ArmeriaMessageDeframer<T> extends HttpDataDeframer<T {
    // subsclass implement like the following 
    List<T> process() {
        if (readableBytes() > 5) { 
            // Read data from buffers
            byte header = readByte();
            int lenght = readInt();
            ... 
        } else {
            // Wait more data
 	        return List.of();
        }
    }
}
```

**Comment by trustin** [comment]
> In `MultipartEncoder` case, it does not need byte-level operations. 

Really? In chunked transfer encoding, wouldn't be a boundary split into multiple chunks although rare?

**Comment by trustin** [comment]
Even if it's not chunked transfer encoding, HTTP decoders do not have a notion of boundaries, so they will be easily split into many pieces, so we need to do some byte-level aggregation.

**Comment by trustin** [comment]
We also need to try to reuse the `HttpData` in an input stream when producing an output stream to reduce unnecessary memory copies. For example, we could forward the input `HttpData` instance as it is when it is a partial content without any other stuff like boundaries and length headers. Could use `WITH_POOLED_OBJECTS` when necessary as well.

**Comment by trustin** [comment]
That being said, simple byte-level streaming abstraction wouldn't meet all the requirements. Maybe better keeping the implementation at object-level and providing some utilities for managing substates for working at byte level?

**Comment by ikhoon** [comment]
> Really? In chunked transfer encoding, wouldn't be a boundary split into multiple chunks although rare?

That's a good point. I thought only files are to be considered in FilePublisher. https://github.com/ikhoon/armeria/commit/a505474fb088c536c9885b400c3414e136c85a4d#diff-cb3400de0d04cdb4c0a6ea2eb79807c2R133-R134. 
But any body parts could be split into multiple chunks.

**Comment by ikhoon** [comment]
Fixed by #2981
